### ISSUE TITLE
How to fine-tune "stsb-xlm-r-multilingual" model with specific task training data

### ISSUE THREAD

**Issue by mattvan83** [UKPLab/sentence-transformers#808]
Hi,

Thanks for this wonderful repository. I am interested by semantic similarity between french sentences only for the moment and found the "stsb-xlm-r-multilingual" model to work pretty well.

I have looked the training overview but I am a bit lost on how to fine tune this model for my specific task. As said, I only have french sentences at the moment. Could you indicate me which format my specific task training data should have and based on which script should I start for fine-tuning ?

Thanks !

**Comment by nreimers** [comment]
Have a look here:
https://www.sbert.net/docs/training/overview.html

There are many different options how your training data can look like:
- You can have pairs of sentences with a judgment on the similarity 0...1: Then use cosine similarity loss (see the STS examples)
- Or you have pairs with binary labels 0/1 for e.g. no-paraphrase / paraphrase. Then there are various types of loss functions possible depending how your data looks exactly: https://www.sbert.net/docs/package_reference/losses.html

**Comment by mattvan83** [comment]
Hi @nreimers 

If I look at the training overview, I see that is possible to fine-tune my model by continuing training on other data.

In the particular case of `stsb-xlm-r-multilingual` model, I saw that it was trained on NLI dataset then fine-tuned on STS benchmark. If I apply what is mentioned in "Continue training on other data" on my particular dataset (pairs of french sentences with binary label 0/1), do I have to retrain and fine-tune on both NLI and STS benchmark or do I only need to present my new specific data to train ?

Secondly, in my particular setting of actual french-only framework can I fine-tune the multi-lingual model with only pairs of french sentences (phrase/paraphrase) ?

**Comment by nreimers** [comment]
@mattvan83 
You can just load the model and continue training on your data. You don't need NLI / STS data.

**Comment by Matthieu-Tinycoaching** [comment]
Hi @nreimers I recently heard of the new unsupervised methods to learn sentence embeddings as TSDAE. Is it possible to take advantage of this new approach to fine-tune `stsb-xlm-r-multilingual` model with this low cost method or are they different approaches? If yes, how could I do ?

Thanks!

**Comment by nreimers** [comment]
@Matthieu-Tinycoaching 

We haven't tested yet what happens when you continue training with the unsupervised approaches with one of the pre-trained models.

Not sure if it will work

**Comment by Matthieu-Tinycoaching** [comment]
Hi @nreimers,

Following your previous answer:

> There are many different options how your training data can look like:
> 
> - You can have pairs of sentences with a judgment on the similarity 0...1: Then use cosine similarity loss (see the STS examples)
> - Or you have pairs with binary labels 0/1 for e.g. no-paraphrase / paraphrase. Then there are various types of loss functions possible depending how your data looks exactly: https://www.sbert.net/docs/package_reference/losses.html

1/ If I use the first option (cosine similarity loss) how could I objectively obtain the judgment on the similarity 0...1 for each pair of sentence? Has this measure to be compute automatically or based on personal judgment?

2/ For the second option, the loss functions corresponding seem to be `ContrastiveLoss` or `OnlineContrastiveLoss`, aren't they? How to choose between both loss functions?

**Comment by Matthieu-Tinycoaching** [comment]
@nreimers would you have any advices regarding my questions?

Thanks!

**Comment by nreimers** [comment]
Have a look at the MultipleNegativesRankingLoss. On https://www.nils-reimers.de/research-talks/ I have several talks linked where I give an intro how to train embedding models and what type of data is needed

**Comment by Matthieu-Tinycoaching** [comment]
@nreimers thanks for these references to MultipleNegativesRankingLoss, I will check them for sure.

However, if I check this loss defintion on sbert site, I can see that : "

> This loss expects as input a batch consisting of sentence pairs (a_1, p_1), (a_2, p_2)…, (a_n, p_n) where we assume that (a_i, p_i) are a positive pair and (a_i, p_j) for i!=j a negative pair.

"

This seems to be incompatible with the fact that I often have for a specific sentence `a_1` multiple corresponding positive pairs `p_1, p_2 ... p_n`, where (a_1, p_1), (a_1, p_2) ... (a_1, p_n) are all positive pairs. Is MultipleNegativesRankingLoss adapted in this case? If not, what loss is best adapted for this specific case?

**Comment by Matthieu-Tinycoaching** [comment]
Thanks! 

Could I have a number of mini-batches as needed (max number of positive pairs across all sentences)? Aren't there constraints on number of mini-batches or number of pairs within on mini-batch for fine-tuning?

**Comment by nreimers** [comment]
Don't understand your question. You must create batches with the described properties of the loss

**Comment by Matthieu-Tinycoaching** [comment]
@nreimers sorry for the poor wording.

I just want to be sure that for fine-tuning there's no limits in terms of number or size of batches?

**Comment by nreimers** [comment]
Size of batch is limited by your GPU memory. At some point, it will overflow.

Number of batches is not limited
